bus-mapping/src/evm/opcodes/calldataload.rs

```diff
@@ -0,0 +1,207 @@
+                &mut exec_step,
+                RW::READ,
+                CallContextOp {
+                    call_id: state.call()?.call_id,
+                    field: CallContextField::CallerId,
+                    value: state.call()?.caller_id.into(),
+                },
+            );
+        }
+
+        let (offset, src_addr_end) = (
+            offset.as_usize(),
+            state.call()?.call_data_offset as usize + state.call()?.call_data_length as usize,
+        );
+        let is_root = state.call()?.is_root;
+        let calldata_word = (0..32)
+            .map(|idx| {
+                let addr = offset + idx;
+                if addr < src_addr_end {
+                    if is_root {
+                        state.tx.input[addr]
+                    } else {
+                        // TODO: read caller memory
+                        unimplemented!();
+                    }
+                } else {
+                    0
+                }
+            })
+            .collect::<Vec<u8>>();
+
+        state.push_stack_op(
+            &mut exec_step,
+            RW::WRITE,
+            geth_step.stack.last_filled(),
+            U256::from_big_endian(&calldata_word),
+        )?;
+
+        Ok(vec![exec_step])
+    }
+}
+
+#[cfg(test)]
+mod calldataload_tests {
+    use eth_types::{
+        bytecode,
+        evm_types::{OpcodeId, StackAddress},
+        geth_types::GethData,
+        Word,
+    };
+    use mock::{
+        test_ctx::helpers::{account_0_code_account_1_no_code, tx_from_1_to_0},
+        TestContext,
+    };
+    use rand::random;
+
+    use crate::{circuit_input_builder::ExecState, mock::BlockData, operation::StackOp};
+
+    use super::*;
+
+    fn rand_bytes(size: usize) -> Vec<u8> {
+        (0..size).map(|_| random()).collect::<Vec<u8>>()
+    }
+
+    #[test]
+    fn calldataload_opcode_impl() {
+        // 1. should be right padded
+        test_ok(0u64, vec![1u8, 2u8], {
+            let mut v = vec![0u8; 32];
+            v[0] = 1u8;
+            v[1] = 2u8;
+            Word::from_big_endian(&v)
+        });
+        // 2. exactly 32 bytes
+        let calldata = rand_bytes(32);
+        test_ok(0u64, calldata.clone(), Word::from_big_endian(&calldata));
+        // 3. take only 32 bytes
+        let calldata = rand_bytes(64);
+        test_ok(
+            12u64,
+            calldata.clone(),
+            Word::from_big_endian(&calldata[12..44]),
+        );
+    }
+
+    fn test_ok(offset: u64, calldata: Vec<u8>, calldata_word: Word) {
+        let code = bytecode! {
+            PUSH32(offset)
+            CALLDATALOAD
+            STOP
+        };
+
+        let mut block: GethData = TestContext::<2, 1>::new(
+            None,
+            account_0_code_account_1_no_code(code),
+            tx_from_1_to_0,
+            |block, _tx| block.number(0xcafeu64),
+        )
+        .unwrap()
+        .into();
+        block.eth_block.transactions[0].input = calldata.into();
```

Comment:
Why not adding the input in the `TestContext` builder directly?

Reply:
Yes, that makes sense. The last commit was simply resolving conflicts from recent upstream updates, so I hadn't had the chance to look into the `TestContext`. Will refactor 👍